### PR TITLE
[BEAM-XXX] Port join extensions to Python

### PR DESCRIPTION
**Pull request by abergmeier**
This is an attempt to port join extensions (Java SDK) to Python.

---------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by abergmeier** [comment]
R: @aaltay
R: @charlesccychen

**Comment by abergmeier** [comment]
This is a first naive code attempt. For the following tasks I could really use some guidance with the following questions:
1. How to handle Tags.
   In Java, the code uses internal knowledge to create what seems to be a class identity? There seem to be some Tags being used in Python but it doesn't seem to have the same power.
   Would be great if someone could shed some light why Tags are there in the first place and how the strategy in Python is.

2. How to handle KV
   I have a hard time finding examples in code that use KV. Since there is a typehint I used this as far as possible in Python code. For the actual KV I then used tuples. Is this fine?

3. Coders
   There seems to be some Coder support but it seems like there is no equivalent to `PCollection.setCoder`. How is this supposed to work in Python?

4. CoGbk*
There seem to be no further util classes around `CoGbk`*. I assume for the most part will have to implement these.

File: sdks/python/apache_beam/transforms/join/keyedpcollectiontuple.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import apache_beam as beam
+from apache_beam.typehints import typehints
+from apache_beam.coders.coders import Coder
+
+K = TypeVar("K")
+V = TypeVar("V")
+InputT = TypeVar("InputT")
+OutputT = TypeVar("OutputT")
+
+class KeyedPCollectionTuple(Generic[K]):
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/transforms/join/keyedpcollectiontuple.py:12]: I think it should be possible to add the Join library to Python without porting `KeyedPCollectionTuple`. In Java this is necessary for CoGroupByKey, but in Python we represent tagged/keyed PCollections as `Dict[str, PCollection]`, you can see how this works with CoGroupByKey in that transform's API docs: https://beam.apache.org/releases/pydoc/2.28.0/apache_beam.transforms.util.html#apache_beam.transforms.util.CoGroupByKey

**Comment by abergmeier** [comment]
> [ ] Please create an account at issues.apache.org/jira so we can file a jira and assign it to you

Done. Same username.

**Comment by abergmeier** [comment]
> [ ] I think it would make sense to put everything for this effort in apache_beam.transforms.join, since we don't split out extensions for the Python SDK.

Done.

**Comment by TheNeuralBit** [comment]
Sorry I completely missed the questions you asked!

> 1. How to handle Tags.
~~In Java, the code uses internal knowledge to create what seems to be a class identity? There seem to be some Tags being used in Python but it doesn't seem to have the same power.
Would be great if someone could shed some light why Tags are there in the first place and how the strategy in Python is.~~
Since this produces a simpler dict in Python, will hardcode the strings.

:+1: 

> 2. How to handle KV
I have a hard time finding examples in code that use KV. Since there is a typehint I used this as far as possible in Python code. For the actual KV I then used tuples. Is this fine?

Yes 2-tuples are the preferred way to represent KVs at execution time in Python.

> 3. Coders
There seems to be some Coder support but it seems like there is no equivalent to PCollection.setCoder. How is this supposed to work in Python?

In Python we always infer the coder from the PCollection's element_type, which is determined from typehints. Often we just fall back to FastPrimitivesCoder. In your case it should be enough to just use `typehints.KV`.

> 4. CoGbk*
There seem to be no further util classes around CoGbk*. I assume for the most part will have to implement these.

As discussed elsewhere, we generally just use python primitives rather than the CoGbk* util classes. Let me know if there are specific ones that don't seem to be supported in Python.

**Comment by TheNeuralBit** [comment]
> > [ ] Please create an account at issues.apache.org/jira so we can file a jira and assign it to you
> Done. Same username.

@aaltay could you add @abergmeier as a contributor in jira?

**Comment by abergmeier** [comment]
> In Python we always infer the coder from the PCollection's element_type, which is determined from typehints. Often we just fall back to FastPrimitivesCoder.

TBH infering/relying on (implicit) typehints sounds scary to me. Will give this a try and report back.

> In your case it should be enough to just use typehints.KV.

I hope they are correct now. For now I tried fully qualifying all typehints.

> As discussed elsewhere, we generally just use python primitives rather than the CoGbk* util classes. Let me know if there are specific ones that don't seem to be supported in Python.

Yes thanks, indeed for now it seems I got rid of all of these.

**Comment by abergmeier** [comment]
Could anyone point to where

> (left, right)
      | "Do me" >> LeftOuterJoin(nullValue=None)

could be enabled? I saw that this is done somehow for:

> Dict | str >> CoGroupByKey()

But I am not sure how the `|` gets enabled by code.

**Comment by aaltay** [comment]
> > > [ ] Please create an account at issues.apache.org/jira so we can file a jira and assign it to you
> > > Done. Same username.
> 
> @aaltay could you add @abergmeier as a contributor in jira?

Done. Sorry for the delay.

Added both abergmeier and abergmeieratotto. Looks like both are your accounts.

**Comment by TheNeuralBit** [comment]
Sorry for not getting back sooner. PTransform has implementations for the operators: :https://github.com/apache/beam/blob/2306bd0733711f8c7807d059256f9c5b0432f5f7/sdks/python/apache_beam/transforms/ptransform.py#L539

It just passes "pvalueish" directly to the PTransform's `expand` implementation. I don't think it pulls apart tuples or dicts, they're just passed through directly, so you'll need to handle them as CoGroupByKey does: https://github.com/apache/beam/blob/2306bd0733711f8c7807d059256f9c5b0432f5f7/sdks/python/apache_beam/transforms/util.py#L190-L202

It would be worth generalizing this somehow.

**Comment by aaltay** [comment]
@abergmeier - What would be the next step on this PR?

**Comment by abergmeier** [comment]
> What would be the next step on this PR?

We decided to evolve the code in our internal app first (thus no public code changes for now).

We tested out various API designs and landed on one that we internally find pleasing.

Now we started writing tests to ensure various aspects are working and coverage is there.
In parallel we ensured that the code is already quite well type annotated.

Especially with type annotation we now found several aspects to be lacking in decorator code.
I am currently trying to understand what magic beam is trying to apply and why some of our cases don't work out.

The thing that currently is most annoying to me is that there is no public chat to have a quick talk about things like design decision in Beam. The whole development process seems a bit heavy handed :(

**Comment by aaltay** [comment]
> > What would be the next step on this PR?
> 
> We decided to evolve the code in our internal app first (thus no public code changes for now).
> 
> We tested out various API designs and landed on one that we internally find pleasing.
> 
> Now we started writing tests to ensure various aspects are working and coverage is there.
> In parallel we ensured that the code is already quite well type annotated.
> 
> Especially with type annotation we now found several aspects to be lacking in decorator code.
> I am currently trying to understand what magic beam is trying to apply and why some of our cases don't work out.
> 
> The thing that currently is most annoying to me is that there is no public chat to have a quick talk about things like design decision in Beam. The whole development process seems a bit heavy handed :(

I understand.

We do have a slack channel, but not all relevant people are online there all the time. You should be able to schedule a time with relevant folks to have a quick chat on slack though.

**Comment by abergmeier** [comment]
> We do have a slack channel

I was under the impression that only "official" devs can join that. But perhaps I missed something...

**Comment by aaltay** [comment]
> > We do have a slack channel
> 
> I was under the impression that only "official" devs can join that. But perhaps I missed something...

Anyone can join. See slack information on this page: https://beam.apache.org/contribute/

(If this is not clear, we can clarify the documents.)

**Comment by abergmeier** [comment]
So our code diverged from something general so much that I won't make sense for general use case.
Not even sure this was a good idea to begin with - at least for GCP.
Thus will close this now.

**Comment by TheNeuralBit** [comment]
> So our code diverged from something general so much that I won't make sense for general use case.
> Not even sure this was a good idea to begin with - at least for GCP.
> Thus will close this now.

Ack, thank you for following up on this. If you feel anything you've built would make sense to generalize a contribution is always welcome :)

**Pull request closed without merging**